plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java

```diff
@@ -0,0 +1,376 @@
+    {
+        return this.includeQueryTypes.stream().map(Optional::of).toList().contains(queryContext.getQueryType());
+    }
+
+    private UUID getQueryId(QueryMetadata queryMetadata)
+    {
+        return UUID.nameUUIDFromBytes(queryMetadata.getQueryId().getBytes(StandardCharsets.UTF_8));
+    }
+
+    private OpenLineage.RunFacet getTrinoQueryContextFacet(QueryContext queryContext)
+    {
+        OpenLineage.RunFacet queryContextFacet = openLineage.newRunFacet();
+
+        queryContextFacet
+                .getAdditionalProperties()
+                .put("serverAddress", queryContext.getServerAddress());
+        queryContextFacet
+                .getAdditionalProperties()
+                .put("serverVersion", queryContext.getServerVersion());
+        queryContextFacet
+                .getAdditionalProperties()
+                .put("environment", queryContext.getEnvironment());
+        queryContext.getQueryType().ifPresent(queryType ->
+                queryContextFacet
+                        .getAdditionalProperties()
+                        .put("queryType", queryType));
+
+        return queryContextFacet;
+    }
+
+    private OpenLineage.RunFacet getTrinoMetadataFacet(QueryMetadata queryMetadata)
+    {
+        OpenLineage.RunFacet trinoMetadataFacet = openLineage.newRunFacet();
+
+        queryMetadata.getPlan().ifPresent(
+                queryPlan -> trinoMetadataFacet
+                        .getAdditionalProperties()
+                        .put("queryPlan", queryPlan));
+
+        queryMetadata.getTransactionId().ifPresent(
+                transactionId -> trinoMetadataFacet
+                        .getAdditionalProperties()
+                        .put("transactionId", transactionId));
+
+        return trinoMetadataFacet;
+    }
+
+    private OpenLineage.RunFacet getTrinoQueryStatisticsFacet(QueryStatistics queryStatistics)
+            throws IllegalAccessException
+    {
+        OpenLineage.RunFacet trinoQueryStatisticsFacet = openLineage.newRunFacet();
+
+        for (Field field : queryStatistics.getClass().getDeclaredFields()) {
+            field.setAccessible(true);
+            trinoQueryStatisticsFacet
+                    .getAdditionalProperties()
+                    .put(field.getName(), String.valueOf(field.get(queryStatistics)));
+        }
+        return trinoQueryStatisticsFacet;
+    }
+
+    public OpenLineage.RunEvent getStartEvent(UUID runID, QueryCreatedEvent queryCreatedEvent)
+            throws IllegalAccessException
+    {
+        OpenLineage.RunFacetsBuilder runFacetsBuilder = openLineage.newRunFacetsBuilder();
+        runFacetsBuilder.put(OpenLineageTrinoFacet.TRINO_METADATA.getText(),
+                getTrinoMetadataFacet(queryCreatedEvent.getMetadata()));
+        runFacetsBuilder.put(OpenLineageTrinoFacet.TRINO_QUERY_CONTEXT.getText(),
+                getTrinoQueryContextFacet(queryCreatedEvent.getContext()));
+
+        OpenLineage.RunEvent startEvent =
```

Comment:
Can we have return statement here ? i.e return openLineage.newRunEventBuilder()`